js/modules/k6/experimental/csv/module.go

```diff
@@ -0,0 +1,289 @@
+		reject(fmt.Errorf("first arg doesn't appear to be a *file.File instance"))
+		return promise
+	}
+
+	parserOptions := newDefaultParserOptions()
+	if options != nil {
+		var err error
+		parserOptions, err = newParserOptionsFrom(options.ToObject(rt))
+		if err != nil {
+			reject(fmt.Errorf("encountered an error while interpreting Parser options; reason: %w", err))
+			return promise
+		}
+	}
+
+	r, err := NewReaderFrom(fileObj.ReadSeekStater, parserOptions)
+	if err != nil {
+		reject(fmt.Errorf("failed to create a new parser; reason: %w", err))
+		return promise
+	}
+
+	go func() {
+		underlyingSharedArrayName := parseSharedArrayNamePrefix + strconv.Itoa(time.Now().Nanosecond())
+
+		// Because we rely on the data module to create the shared array, we need to
+		// make sure that the data module is initialized before we can proceed, and that we don't instantiate
+		// it multiple times.
+		//
+		// As such we hold a single instance of it in the RootModule, and we use it to create the shared array.
+		resolve(mi.RootModule.dataModuleInstance.NewSharedArrayFrom(mi.vu.Runtime(), underlyingSharedArrayName, r))
+	}()
+
+	return promise
+}
+
+// NewParser creates a new CSV parser instance.
+func (mi *ModuleInstance) NewParser(call sobek.ConstructorCall) *sobek.Object {
+	rt := mi.vu.Runtime()
+
+	if mi.vu.State() != nil {
+		common.Throw(rt, errors.New("csv Parser constructor must be called in the init context"))
+	}
+
+	if len(call.Arguments) < 1 || sobek.IsUndefined(call.Argument(0)) {
+		common.Throw(rt, fmt.Errorf("parser constructor takes at least one non-nil source argument"))
+	}
+
+	// 1. Make sure the Goja object is a fs.File (sobek operation)
+	var file fs.File
+	if err := mi.vu.Runtime().ExportTo(call.Argument(0), &file); err != nil {
+		common.Throw(mi.vu.Runtime(), fmt.Errorf("first arg doesn't appear to be a *file.File instance"))
+	}
+
+	options := newDefaultParserOptions()
+	if len(call.Arguments) == 2 && !sobek.IsUndefined(call.Argument(1)) {
+		var err error
+		options, err = newParserOptionsFrom(call.Argument(1).ToObject(rt))
+		if err != nil {
+			common.Throw(rt, fmt.Errorf("encountered an error while interpreting Parser options; reason: %w", err))
+		}
+	}
+
+	// Instantiate and configure a csv reader using the provided file and options
+	r, err := NewReaderFrom(file.ReadSeekStater, options)
+	if err != nil {
+		common.Throw(rt, fmt.Errorf("failed to create a new parser; reason: %w", err))
+	}
+
+	// Create a new Parser instance
+	parser := Parser{
+		reader:  r,
+		options: options,
+		vu:      mi.vu,
+	}
+
+	return rt.ToValue(&parser).ToObject(rt)
+}
+
+// Next returns the next row in the CSV file.
+func (p *Parser) Next() *sobek.Promise {
+	promise, resolve, reject := promises.New(p.vu)
+
+	go func() {
+		var records []string
+		var done bool
+		var err error
+
+		records, err = p.reader.Read()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				done = true
+			} else {
+				reject(err)
+				return
+			}
```

Comment:
Doesn't that suggestion modify the behavior? 🤔 

Like, when `errors.Is(err, io.EOF) == true`, the promise will be rejected with the error, instead of being resolved with done.

If we want to simplify it, I think it should be something like:

```
if err != nil {
	if !errors.Is(err, io.EOF) {
		reject(err)
		return
	}
	done = true
}
```

Reply:
My bad! You're right, so then I'm resolving the tread, but @oleiade feel free to incorporate the @joanlopez 's proposal